File: sdks/python/examples/basic_usage.py
"""Basic store + search + forget flow.

Run with:

    export ENGRAM_API_KEY=pr_live_...
    python examples/basic_usage.py

Prints the stored id, every search hit, and the forget outcome. Every
network call is wrapped in a single ``EngramError`` handler so the
example doubles as a template for real error handling — replace the
``print`` calls with your own logging.
"""

from __future__ import annotations

import sys

from engram import (
    EngramAuthError,
    EngramClient,
    EngramError,
)


def main() -> int:
    try:
        client = EngramClient()
    except EngramError as exc:
        # Usually "no api key set". Exit with a clear message instead
        # of a traceback so a CI operator can see what's wrong.
        print("Engram client setup failed: {}".format(exc), file=sys.stderr)
        return 2

    try:
        # 1. Store
        stored = client.store(
            "Production Postgres runs on port 5433 with pgvector 0.7.0",
            category="infra",
            importance=0.9,
            metadata={"source": "example", "owner": "platform-team"},
        )
        print("[store]   id={} status={}".format(stored.id, stored.status))

        # 2. Search
        hits = client.search(
            "what port does prod postgres use",
            top_k=5,
        )
        print("[search]  {} results (tokens={})".format(len(hits.results), hits.query_tokens))
        for i, hit in enumerate(hits.results, start=1):
            print("  {}. [{}] {:.2f}  {}".format(i, hit.tier or "?", hit.score, hit.text[:80]))

        # 3. Forget the one we just stored
        gone = client.forget(stored.id)
        print("[forget]  status={} found={}".format(gone.status, gone.found))

        # 4. Health probe — last because it's the cheapest check
        health = client.health()
        print("[health]  api={} qdrant={} version={}".format(
            health.api, health.qdrant, health.version,
        ))
    except EngramAuthError as exc:
        print("Auth failed — is ENGRAM_API_KEY current? ({})".format(exc), file=sys.stderr)
        return 1
    except EngramError as exc:
        print("Engram call failed: {}".format(exc), file=sys.stderr)
        return 1
    finally:
        client.close()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: sdks/typescript/examples/basic-usage.ts
/**
 * Basic usage: store a memory, search for it, forget it.
 *
 * Run with an API key in your environment:
 *   ENGRAM_API_KEY=pr_live_... npx tsx examples/basic-usage.ts
 *
 * No env-var magic inside the SDK — we read it here at the edge and
 * pass it explicitly. That way the SDK stays pure and this script
 * stays obvious.
 */

import { EngramClient, EngramAuthError, EngramRateLimitError } from "../src/index.js";

async function main(): Promise<void> {
  const apiKey = process.env.ENGRAM_API_KEY;
  if (!apiKey) {
    console.error("Set ENGRAM_API_KEY first.");
    process.exit(1);
  }

  const client = new EngramClient({ apiKey });

  try {
    // 1. Store a memory
    const stored = await client.store({
      text: "Production database is PostgreSQL 16 on Neon, us-east-1.",
      category: "infrastructure",
      importance: 0.9,
      metadata: { source: "architecture-doc", verified: true },
    });
    console.log("Stored:", stored.id, "->", stored.category);

    // 2. Search for it. `topK` maps to the wire-level `limit`.
    const hits = await client.search({
      query: "what database does production use",
      topK: 3,
    });
    console.log(`Got ${hits.results.length} results (${hits.query_tokens} query tokens):`);
    for (const r of hits.results) {
      console.log(` - [${r.score.toFixed(3)}] ${r.text}`);
    }

    // 3. Forget it
    if (hits.results[0]) {
      const gone = await client.forget(hits.results[0].id);
      console.log("Forgot:", gone.status, gone.id);
    }
  } catch (err) {
    // Error classification lets you branch on cause rather than
    // substring-matching error messages.
    if (err instanceof EngramAuthError) {
      console.error("Auth failed — check your API key.");
      process.exit(2);
    }
    if (err instanceof EngramRateLimitError) {
      console.error(
        `Rate limited. Retry after ${err.retryAfter ?? "?"}s. Upgrade at https://app.engrammemory.ai/dashboard/billing`,
      );
      process.exit(3);
    }
    throw err;
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
